types/context_test.go

```diff
@@ -98,3 +98,82 @@ func TestLogContext(t *testing.T) {
 	ctx.Logger().Error("error")
 	require.Equal(t, *logger.logs, []string{"debug", "info", "error"})
 }
+
+type dummy int64
+
+func (d dummy) Clone() interface{} {
+	return d
+}
+
+func TestWithPrimitive(t *testing.T) {
+	ctx := types.NewContext(nil, abci.Header{}, false, log.NewNopLogger())
+
+	clonerkey := "cloner"
+	stringkey := "string"
+	int32key := "int32"
+	uint32key := "uint32"
+	uint64key := "uint64"
+
+	keys := []string{clonerkey, stringkey, int32key, uint32key, uint64key}
+
+	for _, key := range keys {
+		require.Nil(t, ctx.Value(key))
+	}
+
+	clonerval := dummy(1)
+	stringval := "string"
+	int32val := int32(1)
+	uint32val := uint32(2)
+	uint64val := uint64(3)
+
+	ctx = ctx.
+		WithCloner(clonerkey, clonerval).
+		WithString(stringkey, stringval).
+		WithInt32(int32key, int32val).
+		WithUint32(uint32key, uint32val).
+		WithUint64(uint64key, uint64val)
+
+	require.Equal(t, clonerval, ctx.Value(clonerkey))
+	require.Equal(t, stringval, ctx.Value(stringkey))
+	require.Equal(t, int32val, ctx.Value(int32key))
+	require.Equal(t, uint32val, ctx.Value(uint32key))
+	require.Equal(t, uint64val, ctx.Value(uint64key))
+}
+
+func TestWithCustom(t *testing.T) {
```

Comment:
Can you add a couple of comments to this. It takes a bit of parsing to figure out whats going on in this test.